src/pymor/algorithms/timestepping.py

```diff
@@ -10,44 +10,363 @@
 by an external library into an instationary |Model|.
 
 Currently, implementations of :func:`explicit_euler` and :func:`implicit_euler`
-time-stepping are provided. The :class:`TimeStepper` defines a
-common interface that has to be fulfilled by the time-steppers used
-by |InstationaryModel|. The classes :class:`ExplicitEulerTimeStepper`
-and :class:`ImplicitEulerTimeStepper` encapsulate :func:`explicit_euler` and
-:func:`implicit_euler` to provide this interface.
+time-stepping are provided, based on the :class:`ExplicitEulerTimeStepper` and
+:class:`ImplicitEulerTimeStepper` classes. These derive from :class:`TimeStepper`,
+which defines a common interface that has to be fulfilled by the time-steppers used
+by |InstationaryModel|.
 """
 
+from numbers import Number
+import numpy as np
+
 from pymor.core.base import ImmutableObject, abstractmethod
+from pymor.operators.constructions import IdentityOperator, VectorArrayOperator, ZeroOperator
 from pymor.operators.interface import Operator
+from pymor.parameters.base import Mu
+from pymor.tools import floatcmp
 from pymor.vectorarrays.interface import VectorArray
 
 
 class TimeStepper(ImmutableObject):
     """Interface for time-stepping algorithms.
 
-    Algorithms implementing this interface solve time-dependent problems
+    Algorithms implementing this interface solve time-dependent initial value problems
     of the form ::
 
-        M * d_t u + A(u, mu, t) = F(mu, t).
+        M * d_t u + A(u, mu, t) = F(mu, t),
+                     u(mu, t_0) = u_0(mu).
+
+    Time-steppers used by |InstationaryModel| have to fulfill this interface. Time evolution can either be performed
+    by calling :meth:`solve`, or by manually calling :meth:`bootstrap` and :meth:`step`.
 
-    Time-steppers used by |InstationaryModel| have to fulfill
-    this interface.
+    Note that :meth:`bootstrap` and :meth:`step` are designed to support a minimal implementation of :meth:`solve`,
+    as the outer loop in :meth:`solve` may arise manually in several places.
+
+    Parameters
+    ----------
+    initial_time
+        The time at which to begin time-stepping.
+    end_time
+        The time until which to perform time-stepping.
+    num_values
+        The number of returned vectors of the solution trajectory. If `None`, each intermediate vector that is
+        calculated is returned. Else an interpolation of the calculated vectors on an equidistant temporal grid is
+        returned, using an appropriate interpolation of the respective time stepper.
+
+    Attributes
+    ----------
+    steps
+        The number of previous steps required to determine the next. I.e. steps == 1 for a single-step method and
+        steps > 1 for a multi-step method.
     """
 
+    initial_time = None
+    end_time = None
+    steps = None
+    implicit = False
+    num_values = None
+
+    def __init__(self, initial_time, end_time, num_values=None):
```

Comment:
I would be open to move `num_values` to `__init__`. However, this would mean, that you need a new model whenever you want to change this value. What would be the argument in favor of making this an attribute of the time stepper?

Reply:
It is similar to the current situation, where we also need a new model if we change `nt`. I do not have a strong opinion on that, apart from the fact that the one being responsible for the interpolation should hold the `num_values` (see other reply on that).